### PR TITLE
Add function splice and tee

### PR DESCRIPTION
**Pull request by NobodyXu**
The syscalls [`splice`] and [`tee`] can copy data between pipes/pipe and other fd efficiently without copying data between userspace and kernel.

This PR add
 - `AtomicLen`
 - `tee` for copying data between pipes without consuming it
 - `tee_atomic` for copying data between pipes without consuming it atomically
 - `splice` for copying data between pipes
 - `splice_atomic` for copying data between pipes atomically
 - `PipeRead::splice_to` for copying data from pipe
 - `PipeWrite::splice_from` for copying data into pipe
 - `PipeWrite::splice_from_atomic` for copying data into pipe atomically

This PR also added several tests for the newly added `tee` and `splice`.

This PR also mark `AtomicWriteBuffer::new` as `const`.

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

[`splice`]: https://man7.org/linux/man-pages/man2/splice.2.html
[`tee`]: https://man7.org/linux/man-pages/man2/tee.2.html

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -26,11 +26,15 @@ use std::io;
 use std::mem;
 use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
 use std::pin::Pin;
+use std::ptr;
```
Comment (yskszk63) [inline src/lib.rs:29]: Can you please fix the `unused_imports` warning on non-Linux?

```suggestion
#[cfg(target_os = "linux")]
use std::ptr;
```

Comment (NobodyXu) [reply src/lib.rs:30]: Thanks, I foegot to add that to str::ptr

**Comment by yskszk63** [comment]
@NobodyXu 
Sorry. #14 merge caused conflicts.
Can you resolve the conflict?

```rust
/// A buffer that can be written atomically
<<<<<<< master
#[derive(Debug)]
=======
#[derive(Copy, Clone, Debug)]
>>>>>>> master
pub struct AtomicWriteBuffer<'a>(&'a [u8]);
```

**Comment by NobodyXu** [comment]
@yskszk63 Done.

**Comment by yskszk63** [comment]
Thank you!

**Comment by yskszk63** [comment]
It has been released as v0.2.4.

**Pull request merged**